### PR TITLE
fixed trainer tr_loss memory leak

### LINKED ISSUE #6939: PyTorch (with GPU) Trainer leaks CPU memory on Google Colab
## Environment info
- `transformers` version: 3.1.0
- Platform: Linux-4.19.112+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.6.9
- PyTorch version (GPU?): 1.6.0+cu101 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help
Trainer: @sgugger

## Information

Model I am using `RoBERTaForMaskedLM`:
`
config = RobertaConfig(
    vocab_size=32_000,
    max_position_embeddings=256+2,
    num_attention_heads=12,
    num_hidden_layers=6,
    type_vocab_size=1,
)
`
The problem arises when using:
* [ ] [my own modified scripts](https://colab.research.google.com/drive/1bVv6V9IIhTNXvWbpUvQyPyRpK3LOgFfz?usp=sharing): I am trying to follow the official blog tutorial on training a language model from scratch. I have made a few changes from the official script (using Marathi OSCAR corpus, changed model config, and vocabulary size, and am fetching the sentences in the dataset on the fly

The tasks I am working on is:
* [ ] my own task or dataset: Masked Language Modelling with RoBERTa on Marathi Oscar Corpus

## To reproduce

1. Run this [colab notebook](https://colab.research.google.com/drive/1bVv6V9IIhTNXvWbpUvQyPyRpK3LOgFfz?usp=sharing).

## Expected behavior
The RAM consumption starts rising towards the end of the first epoch, ultimately crashing the entire session due to full memory consumption (12.72GB of Colab RAM)

### PR DESCRIPTION
**Pull request by StuartMesham**
Fixes #6939 

The Trainer class contains the memory leak described [here](https://discuss.pytorch.org/t/cpu-ram-usage-increasing-for-every-epoch/24475/6). It is not specific to any particular model type and will occur with any model trained using the trainer class.

It is fixed in this pull request.

The issue is demonstrated in [this Colab notebook](https://colab.research.google.com/drive/1KQZEiZtfY14sDiAQnfw0gkKj9eC6XpTm?usp=sharing). The model trains for around 20 000 steps (this takes around 10 minutes on a t4) before using up all 12GB of RAM.

### CODE REVIEW

**Comment by julien-c** [comment]
I think this was touched by @jysohn23 recently so pinging him here

**Comment by sgugger** [comment]
I think the problem is not that we don't call `.item()` but that we don't call `.detach()`, which means some variables are kept forever for a backward pass (that is never called).
The `.item()` are removed because it's needed for faster TPU training.

**Comment by jysohn23** [comment]
This ends up calling `.item()` every step which ends up hurting performance by like 2X. What @sgugger sounds promising. Can we try that out instead?

**Review by LysandreJik — APPROVED** [review]
~I think this change makes sense from a memory perspective. What do you think @sgugger ?~ My bad, didn't see the latest comments.

**Comment by StuartMesham** [comment]
I have updated it to create a detached tensor instead of calling item().

**Comment by sgugger** [comment]
Can you confirm it fixes the memory leak? This is the right fix IMO (@LysandreJik this might be the fix for the TPU memory leak we has in another issue too.)

**Comment by StuartMesham** [comment]
> Can you confirm it fixes the memory leak? This is the right fix IMO (@LysandreJik this might be the fix for the TPU memory leak we have in another issue too.)

Great! Yes, I have tested to make sure that this fixes the leak.
